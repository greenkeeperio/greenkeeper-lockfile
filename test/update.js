'use strict'

const stub = require('sinon').stub
const childProcess = require('child_process')
const exec = stub(childProcess, 'execSync')
const process = require('process')
const path = require('path')

const update = require('../update')

process.env.JENKINS_URL = 'true'
process.env.BUILD_NUMBER = '1'
process.env.GIT_BRANCH = 'fooo/greenkeeper/my-dependency-1.0.0'

afterAll(() => {
  childProcess.execSync.restore()
})

function testFixture (fixtureDirectory, noLog) {
  exec.reset()
  exec.withArgs('git status --porcelain').returns('1')
  exec.withArgs('git config remote.origin.url').returns('git@github.com:greenkeeperio/greenkeeper-lockfile.git')
  exec.withArgs('npm --version').returns('3.0.0')
  exec.withArgs('npm5 -v').throws()
  if (!noLog) exec.withArgs(`git show --oneline --name-only origin/greenkeeper/my-dependency-1.0.0...master | grep -E '(package-lock.json|npm-shrinkwrap.json|yarn.lock|shrinkwrap.yaml)'`).throws({status: 1, stdout: '', stderr: ''})

  process.chdir(path.join(__dirname, fixtureDirectory))

  let oldGhToken = false
  if (process.env.GH_TOKEN) {
    oldGhToken = process.env.GH_TOKEN
    delete process.env.GH_TOKEN
  }

  update()

  if (oldGhToken) {
    process.env.GH_TOKEN = oldGhToken
  }

  expect(exec.getCalls().map(call => call.args[0])).toMatchSnapshot()
}

const tests = {
  'monorepo: root package': 'fixtures/root-package',
  'monorepo: root and one sub package': 'fixtures/root-and-one-sub',
  'monorepo: root and two sub package': 'fixtures/root-and-two-sub',
  'monorepo: root and two sub package at different levels': 'fixtures/root-and-two-diff-sub',
  'monorepo: no root and one sub package': 'fixtures/no-root-and-one-sub',
  'monorepo: no root and two sub package': 'fixtures/no-root-and-two-sub',
  'monorepo: no root and two sub package at different levels': 'fixtures/no-root-and-two-diff-sub'
}

for (const name in tests) {
  test(name, () => testFixture(tests[name]))
}

// special case
test('monorepo: no package.json', () => {
  testFixture('fixtures/no-package', true)
})
