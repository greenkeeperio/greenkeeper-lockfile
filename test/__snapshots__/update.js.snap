// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`monorepo: no package.json 1`] = `
Array [
  "git remote add gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git || git remote set-url gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git",
  "git config --replace-all remote.origin.fetch +refs/heads/*:refs/remotes/origin/*",
  "git fetch gk-origin",
  "git stash",
  "git checkout master",
  "git show --oneline --name-only origin/greenkeeper/my-dependency-1.0.0...master | grep -E '(package-lock.json|npm-shrinkwrap.json|yarn.lock|shrinkwrap.yaml)'",
  "git checkout -",
]
`;

exports[`monorepo: no root and one sub package 1`] = `
Array [
  "git remote add gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git || git remote set-url gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git",
  "git config --replace-all remote.origin.fetch +refs/heads/*:refs/remotes/origin/*",
  "git fetch gk-origin",
  "git stash",
  "git checkout master",
  "git show --oneline --name-only origin/greenkeeper/my-dependency-1.0.0...master | grep -E '(package-lock.json|npm-shrinkwrap.json|yarn.lock|shrinkwrap.yaml)'",
  "git checkout -",
  "git stash",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "git config user.email \\"support@greenkeeper.io\\"",
  "git config user.name \\"greenkeeperio-bot\\"",
  "git status --porcelain",
  "git commit -m \\"chore(package): update lockfile

https://npm.im/greenkeeper-lockfile\\"",
]
`;

exports[`monorepo: no root and two sub package 1`] = `
Array [
  "git remote add gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git || git remote set-url gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git",
  "git config --replace-all remote.origin.fetch +refs/heads/*:refs/remotes/origin/*",
  "git fetch gk-origin",
  "git stash",
  "git checkout master",
  "git show --oneline --name-only origin/greenkeeper/my-dependency-1.0.0...master | grep -E '(package-lock.json|npm-shrinkwrap.json|yarn.lock|shrinkwrap.yaml)'",
  "git checkout -",
  "git stash",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "git config user.email \\"support@greenkeeper.io\\"",
  "git config user.name \\"greenkeeperio-bot\\"",
  "git status --porcelain",
  "git commit -m \\"chore(package): update lockfile

https://npm.im/greenkeeper-lockfile\\"",
]
`;

exports[`monorepo: no root and two sub package at different levels 1`] = `
Array [
  "git remote add gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git || git remote set-url gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git",
  "git config --replace-all remote.origin.fetch +refs/heads/*:refs/remotes/origin/*",
  "git fetch gk-origin",
  "git stash",
  "git checkout master",
  "git show --oneline --name-only origin/greenkeeper/my-dependency-1.0.0...master | grep -E '(package-lock.json|npm-shrinkwrap.json|yarn.lock|shrinkwrap.yaml)'",
  "git checkout -",
  "git stash",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "git config user.email \\"support@greenkeeper.io\\"",
  "git config user.name \\"greenkeeperio-bot\\"",
  "git status --porcelain",
  "git commit -m \\"chore(package): update lockfile

https://npm.im/greenkeeper-lockfile\\"",
]
`;

exports[`monorepo: root and one sub package 1`] = `
Array [
  "git remote add gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git || git remote set-url gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git",
  "git config --replace-all remote.origin.fetch +refs/heads/*:refs/remotes/origin/*",
  "git fetch gk-origin",
  "git stash",
  "git checkout master",
  "git show --oneline --name-only origin/greenkeeper/my-dependency-1.0.0...master | grep -E '(package-lock.json|npm-shrinkwrap.json|yarn.lock|shrinkwrap.yaml)'",
  "git checkout -",
  "git stash",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "git config user.email \\"support@greenkeeper.io\\"",
  "git config user.name \\"greenkeeperio-bot\\"",
  "git status --porcelain",
  "git commit -m \\"chore(package): update lockfile

https://npm.im/greenkeeper-lockfile\\"",
]
`;

exports[`monorepo: root and two sub package 1`] = `
Array [
  "git remote add gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git || git remote set-url gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git",
  "git config --replace-all remote.origin.fetch +refs/heads/*:refs/remotes/origin/*",
  "git fetch gk-origin",
  "git stash",
  "git checkout master",
  "git show --oneline --name-only origin/greenkeeper/my-dependency-1.0.0...master | grep -E '(package-lock.json|npm-shrinkwrap.json|yarn.lock|shrinkwrap.yaml)'",
  "git checkout -",
  "git stash",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "git config user.email \\"support@greenkeeper.io\\"",
  "git config user.name \\"greenkeeperio-bot\\"",
  "git status --porcelain",
  "git commit -m \\"chore(package): update lockfile

https://npm.im/greenkeeper-lockfile\\"",
]
`;

exports[`monorepo: root and two sub package at different levels 1`] = `
Array [
  "git remote add gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git || git remote set-url gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git",
  "git config --replace-all remote.origin.fetch +refs/heads/*:refs/remotes/origin/*",
  "git fetch gk-origin",
  "git stash",
  "git checkout master",
  "git show --oneline --name-only origin/greenkeeper/my-dependency-1.0.0...master | grep -E '(package-lock.json|npm-shrinkwrap.json|yarn.lock|shrinkwrap.yaml)'",
  "git checkout -",
  "git stash",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "git config user.email \\"support@greenkeeper.io\\"",
  "git config user.name \\"greenkeeperio-bot\\"",
  "git status --porcelain",
  "git commit -m \\"chore(package): update lockfile

https://npm.im/greenkeeper-lockfile\\"",
]
`;

exports[`monorepo: root package 1`] = `
Array [
  "git config remote.origin.url",
  "git config remote.origin.url",
  "git remote add gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git || git remote set-url gk-origin git@github.com:greenkeeperio/greenkeeper-lockfile.git",
  "git config --replace-all remote.origin.fetch +refs/heads/*:refs/remotes/origin/*",
  "git fetch gk-origin",
  "git stash",
  "git checkout master",
  "git show --oneline --name-only origin/greenkeeper/my-dependency-1.0.0...master | grep -E '(package-lock.json|npm-shrinkwrap.json|yarn.lock|shrinkwrap.yaml)'",
  "git checkout -",
  "git stash",
  "npm --version",
  "npm5 -v",
  "npm install -S my-dependency@1.0.0",
  "git status --porcelain",
  "git add npm-shrinkwrap.json 2>/dev/null || true",
  "git add package-lock.json 2>/dev/null || true",
  "git add yarn.lock 2>/dev/null || true",
  "git add shrinkwrap.yaml 2>/dev/null || true",
  "git config user.email \\"support@greenkeeper.io\\"",
  "git config user.name \\"greenkeeperio-bot\\"",
  "git status --porcelain",
  "git commit -m \\"chore(package): update lockfile

https://npm.im/greenkeeper-lockfile\\"",
]
`;
